store/migration/dev/20221027000000##approval_instance.sql

```diff
@@ -0,0 +1,22 @@
+-- approval_instance stores approval instances of third party applications.
+CREATE TABLE approval_instance ( 
+    id SERIAL PRIMARY KEY,
+    row_status row_status NOT NULL DEFAULT 'NORMAL',
+    created_ts BIGINT NOT NULL DEFAULT extract(epoch from now()),
+    updated_ts BIGINT NOT NULL DEFAULT extract(epoch from now()),
+    issue_id INTEGER NOT NULL REFERENCES issue (id),
+    requester_id INTEGER NOT NULL REFERENCES principal (id),
+    approver_id INTEGER NOT NULL REFERENCES principal (id),
+    type TEXT NOT NULL CHECK (type LIKE 'bb.plugin.application.approval-instance.%'),
```

Comment:
plugin and application are synonyms in some sense. I am still trying to understand what does an approval instance represent

Reply:
The terminology "approval instance" is used by Lark.

The object generated when an employee initiates an approval.
https://open.feishu.cn/document/uAjLw4CM/ukTMukTMukTM/reference/approval-v4/instance/overview-approval-instance

Reply:
1. We don't use "instance' suffix, otherwise, many tables should have that "_instance" as well. And if it's for 3rd party, then the table name is more like "external_approval"

1. We are using "bb.plugin.application.approval-instance", so 
* For lark, what's the type name
* What else could under "bb.plugin.application" namespace
* What's the difference between plugin and application, what should be put directly under plugin and what should be put under plugin.application

Reply:
I'm thinking "plugin" as some sort of general name for 3rd party integration stuff, which includes webhook, bot, application, etc..
Since we have "bb.plugin.webhook", so I call this "bb.plugin.application" likewise.
We could have "bb.plugin.application.bot" in the future for example.

Reply:
>  And if it's for 3rd party, then the table name is more like "external_approval"

SGTM

Reply:
> I'm thinking "plugin" as some sort of general name for 3rd party integration stuff, which includes webhook, bot, application, etc.. Since we have "bb.plugin.webhook", so I call this "bb.plugin.application" likewise. We could have "bb.plugin.application.bot" in the future for example.
> 
> For lark, the type name is "bb.plugin.application.approval-instance.lark".

I see, then we are entering the application market place / App store realm. Then I would call bb.plugin.app.lark instead of bb.plugin.application.approval-instance.lark: 

1. Normally, we avoid abbreviation, but app is so universal (like we say App Store instead of Application store)

2. If we think of building an app store, then the 1st level is application name/provider instead of the feature

Reply:
"app" sounds good.
However, the "approval instance" is meant to be generic and applies to other vendors too (in my previous investigation, this table schema can also be used by wecom). So I put the vendor name last. wdyt.

Reply:
This is already inside approval_instance table, thus having "approval-instance" inside the type is redundant